### ISSUE TITLE
Setup process needs feedback

### ISSUE THREAD

**Issue by csvance** [csvance/FTBot#39]
The process of setting up the bot and training it initially doesn't provide much useful information about the learned knowledge of the bot. Features should be added to provide statistics on what the bot knows to the user.

**Comment by Atropha** [comment]
There's definitely _some_ issue that happens when importing a training set from a text file. import_text_file.py does seem to execute correctly and the bot trains on it upon startup, but the weights that are generated from it end up not producing a response when @-ing the bot in Discord, even with a substantial dataset.

I'm not sure if that's an issue with formatting, the importer requiring something more than just lines plaintext or if it's a problem specific to my data-sets. I need to experiment some with different kinds of input to see what works ad what doesn't.

**Comment by csvance** [comment]
That sounds like the correct way to go about things. The latest input_text_file.py from master now has a --verbose option which prints out ever line it imports which may be helpful for debugging purposes. 

Could you upload the contents of your db and weights folders after importing data and training? The files in the db folder are just sqlite databases that store lines of text which get trained on startup. If it looks like the data made it in there ok, I will take a look at the word database. Regardless, I should be able to pinpoint where the problem is because it sounds like it is something that I am not seeing with my current datasets.

**Comment by Atropha** [comment]
Here you go. It's just a snippet of logs from the Discord channel. I attached the txt file I jsut as well, just in case. How did you generate your TrumpTweet weights from the JSON file that's in the trainingdata folder? import_text_file.py did not like it the way it is.

[ShortestDiscord.txt](https://github.com/csvance/armchair-expert/files/1750624/ShortestDiscord.txt)
[Weights.zip](https://github.com/csvance/armchair-expert/files/1750623/Weights.zip)

**Comment by csvance** [comment]
That specific dataset was imported with a custom script back when I was first developing the automated training system. I did use import_text_file.py to import a 300K line dataset though.

Going to take a look through the data you provided and see if I can't figure out the issue after my physics lab this morning, so expect a reply sometime this afternoon (CST / GMT-6).

**Comment by csvance** [comment]
So, I trained a bot with the imported dataset you provided and i get a similar problem. I debugged it to see what was being fed to training, and basically it looks like there might be some sort of encoding issue with the data to begin with. I can say with absolute certainty that data encoded in UTF-8 is supported. When the data is dumped from the database it doesn't look like UTF-8 at all.

Can you upload the original text file you used to with import_text_file.py? I suspect simply converting the encoding of that file to UTF-8 will fix everything BTW. Maybe I should try to do this automatically in the import script, but at bare minimum this needs to be documented!

**Comment by Atropha** [comment]
I|ll give the UTF-8 encoding a shot. The text file I used I attached to my previous post. Gonna try setting the encoding and report back once it's done training!

**Comment by csvance** [comment]
It looks like the file is UTF-8 at first glance. Going to do a bit more digging.

**Comment by csvance** [comment]
Ok, here is what I initially get trying to import the file on my Linux box:
```
Traceback (most recent call last):
  File "import_text_file.py", line 33, in <module>
    main()
  File "import_text_file.py", line 16, in main
    data = open(args.datafile, 'r').read()
  File "/home/phorous/anaconda3/lib/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

I may open the file with a hex editor and take a look or just try to convert it and dumping invalid character sequences for UTF-8.

**Comment by csvance** [comment]
Ok, so I added a conversion to UTF-8 in the import script that looks like it might work. I am training with your dataset now.

**Comment by csvance** [comment]
Still no luck. Viewing the file in the hex editor it does not look like UTF-8. There is a null byte (0x00) in-between every character.

**Comment by csvance** [comment]
I converted the file and am training with it now: [UTF8ShortestDiscord.txt](https://github.com/csvance/armchair-expert/files/1753482/UTF8ShortestDiscord.txt)

**Comment by Atropha** [comment]
Yeah, I just looked through my files and for some reason it got saved as UTF-16 at some point, which inserted those Null characters and retained them even when saving as UTF-8. I fixed it by saving it as an RTF file and opening THAT in Notepad++ xD

**Comment by csvance** [comment]
I am adding a warning when it detects non UTF-8 characters to hopefully avoid this happening to anyone again. Nothing worse than getting familiar with some new software, unintentionally feeding it misformatted information and getting no error message. This is addressed in 777015591e209cf31acb5c7babc0b861ffcc7106.

BTW, I get better results from the bot with your dataset when I am using the structure-model.h5 weights from the Trump bot. Data from chat programs tends to have an issue where many of the responses are extremely short, resulting in small sentence structures in the bots replies. This is often undesirable when you want a comedic effect. markov.json.zlib contains the bots vocabulary, so your bot won't start using words Trump uses. I may upload some pertained structure weights people can use in the future to help with this.

**Comment by Atropha** [comment]
Interesting idea, using different vocabularies and structure weights! I'll have to keep that possibility in mind.
